README.md

```diff
@@ -1,10 +1,238 @@
+	subgraph BE[Extension Node]
+		B2["Hash(Path+Hash1)"]
+	end
+	subgraph CI[Inner Node]
+		C1["Hash(Hash3+Hash4)"]
+	end
+	subgraph CL[Leaf Node]
+		C2[Hash2]
+	end
+	subgraph DL1[Leaf Node]
+		D1[Hash3]
+	end
+	subgraph DL2[Leaf Node]
+		D2[Hash4]
+	end
+	subgraph EL[Leaf Node]
+		E1[Hash1]
+	end
+	Root-->|0| BE
+	Root-->|1| BI
+	BI-->|0| CL
+	BI-->|1| CI
+	CI-->|0| DL1
+	CI-->|1| DL2
+	BE-->EL
+```
+
+#### Lazy Nodes
+
+When importing a tree via `ImportSparseMerkleTree` the tree will be lazily loaded from the root hash provided. As such the initial tree structure would contain just a single lazy node, until the tree is used and nodes have to be resolved from the database, whose digest is the root hash of the tree.
+
+```mermaid
+graph TD
+	subgraph L[Lazy Node]
+		A[rootHash]
+	end
+	subgraph T[Tree]
+		L
+	end
+```
+
+If we were to resolve just this root node, we could have the following tree structure:
+
+```mermaid
+graph TD
+	subgraph I[Inner Node]
+		A["Hash(Hash1 + Hash2)"]
+	end
+	subgraph L1[Lazy Node]
+		B["Hash1"]
+	end
+	subgraph L2[Lazy Node]
+		C["Hash2"]
+	end
+	subgraph T[Tree]
+		I --> L1
+		I --> L2
+	end
+```
+
+Where `Hash(Hash1 + Hash2)` is the same root hash as the previous example.
+
+## Paths
+
+Paths are **only** stored in two types of nodes: Leaf nodes and Extension nodes.
+
+- Extension nodes contain not only the path they represent but also the path bounds (ie. the start and end of the path they cover).
+- Leaf nodes contain the full path which they represent, as well as the value stored at that path.
+
+Inner nodes do **not** contain a path, as they represent a branch in the tree and not a path. As such their children, _if they are extension nodes or leaf nodes_, will hold a path value.
+
+## Values
+
+By default the SMT will use the `hasher` passed into `NewSparseMerkleTree` to hash both the keys into their paths in the tree, as well as the values. This means the data stored in a leaf node will be the hash of the value, not the value itself.
+
+However, if this is not desired, the two option functions `WithPathHasher` and `WithValueHasher` can be used to change the hashing function used for the keys and values respectively.
+
+If `nil` is passed into `WithValueHasher` functions, it will act as identity hasher and store the values unaltered in the tree.
+
+## Hashers & Digests
+
+When creating a new SMT or importing one a `hasher` is provided, typically this would be `sha256.New()` but could be any hasher implementing the go `hash.Hash` interface. By default this hasher, referred to as the `TreeHasher` will be used on both keys (to create paths) and values (to store). But separate hashers can be passed in via the option functions mentioned above.
+
+Whenever we do an operation on the tree, the `PathHasher` is used to hash the key and return its digest - the path. When we store a value in a leaf node we hash it using the `ValueHasher`. These digests are calculated by writing to the hasher and then calling `Sum(nil)`.
```

Comment:
Just FYI, the reason we call `Sum` is because we are computing the `checksum` of the underlying data - https://en.wikipedia.org/wiki/Checksum.

You might write `and then computing the checksum by calling Sum(nil)`.

Up to you if you want to edit the doc but worth noting.